src/main/java/org/commcare/cases/model/StorageIndexedTreeElementModel.java

```diff
@@ -0,0 +1,151 @@
+    protected String entityId;
+
+    @SuppressWarnings("unused")
+    public StorageIndexedTreeElementModel() {
+        // for serialization
+    }
+
+    public StorageIndexedTreeElementModel(Set<String> indices, TreeElement root) {
+        this.indices = new Vector<>(indices);
+        this.root = root;
+        metaDataFields = buildMetadataFields(this.indices);
+    }
+
+    private static String[] buildMetadataFields(List<String> indices) {
+        String[] escapedIndexList = new String[indices.size()];
+        int i = 0;
+        for (String index : indices) {
+            escapedIndexList[i++] = getColFromEntry(index);
+        }
+        return escapedIndexList;
+    }
+
+    public static String getTableName(String fixtureName) {
+        String cleanedName = fixtureName.replace(":", "_").replace(".", "_").replace("-", "_");
+        return STORAGE_KEY_PREFIX + cleanedName;
+    }
+
+    @Override
+    public String[] getMetaDataFields() {
+        return metaDataFields;
+    }
+
+    @Override
+    public Object getMetaData(String fieldName) {
+        if (fieldName.startsWith(ATTR_PREFIX)) {
+            return root.getAttributeValue(null, getEntryFromCol(fieldName).substring(1));
+        } else if (fieldName.startsWith(ELEM_PREFIX)) {
+            // NOTE PLM: The usage of getChild of '0' below assumes indexes
+            // are only made over entries with multiplicity 0
+            TreeElement child = root.getChild(getEntryFromCol(fieldName), 0);
+            IAnswerData value = child.getValue();
+            if (value == null) {
+                return "";
+            } else {
+                return value.uncast().getString();
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public void setID(int id) {
+        this.recordId = id;
+    }
+
+    @Override
+    public int getID() {
+        return recordId;
+    }
+
+    public String getEntityId() {
+        return entityId;
+    }
+
+    public TreeElement getRoot() {
+        return root;
+    }
+
+    @Override
+    public void readExternal(DataInputStream in, PrototypeFactory pf)
+            throws IOException, DeserializationException {
+        recordId = ExtUtil.readInt(in);
+        entityId = ExtUtil.nullIfEmpty(ExtUtil.readString(in));
+        root = (TreeElement)ExtUtil.read(in, TreeElement.class, pf);
+        indices = (Vector<String>)ExtUtil.read(in, new ExtWrapList(String.class), pf);
+        metaDataFields = buildMetadataFields(indices);
+    }
+
+    @Override
+    public void writeExternal(DataOutputStream out) throws IOException {
+        ExtUtil.writeNumeric(out, recordId);
+        ExtUtil.writeString(out, ExtUtil.emptyIfNull(entityId));
+        ExtUtil.write(out, root);
+        ExtUtil.write(out, new ExtWrapList(indices));
+    }
+
+    /**
+     * Turns a column name into the corresponding attribute or element for the TreeElement
+     */
+    public static String getEntryFromCol(String col) {
```

Comment:
This method and the one below could probably have more descriptive names (which might remove the necessity of the method comments). Something like `getElementOrAttributeFromSqlColumnName()` and the reverse.